### ISSUE TITLE
Unnecessary calculation of gradient of all instances in Adversarial Patch

### ISSUE THREAD

**Issue by WladimirDelaCadena** [IBM/adversarial-robustness-toolbox#96]
In the `generate(self, x, y=None):` of the Adversarial Patch attack the following line calculates the gradient for the whole x vector, which contains ALL training instances:
 `gradients = self.classifier.loss_gradient(patched_images, to_categorical(np.broadcast_to(np.array(self.target), x.shape[0]),                                                                   self.classifier.nb_classes))`
However, afterwards the patch is accumulated only from 0 up to `i_batch`, it means only `gradients[0:i_batch]` is used.
Thus, why to calculate the gradients for all instances when they are not used?. I had some _Exhausted Resources Errors_ because the tensor was too big.

**Comment by beat-buesser** [comment]
Hi @WladimirDelaCadena Thank you bringing this to our attention. 

We are very interested in improving the performance of ART and will try to look at it as soon as possible. It is possible that the `AdversarialPatch` code has been only tested with a small `x` so far.

Did you already try to fix it by adapting the size of `x` accordingly?

**Comment by WladimirDelaCadena** [comment]
Well, I set batch_size to the total length of my training set. But that does not scale with many instances. Indeed now it works, but I think the function is not properly implemented since the gradients should be calculated among all training instances. I am doing a loop iterating over the batches, to calculate the gradients for all and accumulate in the patch without scaling problems. But I do not if that is the final solution

**Comment by beat-buesser** [comment]
I agree with your observation, it looks like in the current version there is the implicit assumption that `x` has the size of `batch_size`. I think a loop breaking up x into separate batches might be required here.
Would you be willing to share your solution once it's final?

**Comment by WladimirDelaCadena** [comment]
Yes, I will share it on this issue as soon as I have it

**Comment by beat-buesser** [comment]
That's great! Do you mind if I assign this issue to you?

**Comment by WladimirDelaCadena** [comment]
No problem

**Comment by WladimirDelaCadena** [comment]
I used this for my solution, I hope it can help 

```python
step_size = int(x.shape[0]/self.batch_size)
for k in range(step_size):
    rlow = k*self.batch_size
    rhigh =  k*self.batch_size+self.batch_size
    gradients = self.classifier.loss_gradient(patched_images[rlow:rhigh],y[rlow:rhigh])
    patch_gradients = np.zeros_like(self.patch)
    for i_batch in range(self.batch_size):
        patch_gradients_i = self._reverse_transformation(gradients[i_batch, :, :, :], patch_mask_transformed[i_batch, :, :, :],transforms[i_batch])
        patch_gradients += patch_gradients_i           
    patch_gradients = patch_gradients / self.batch_size
    self.patch -= patch_gradients * self.learning_rate

```

**Comment by beat-buesser** [comment]
That's great! Thank you very much for sharing your solution.

Would you like to make a commit and pull request to merge your solution into our `dev` branch?
